[Resubmit][S372460 follow up] Reduce embedding feature validation failure carry-on impact

dlmetamovefast opened the pull request:
Summary:
## Context
The embedding feature validation for GatherRangeToDense was added in the previous diff: D18031155. The logic will check the mismatchedranges or empty ranges in the whole model lifecycle, once the ratio exceeds some thresehold, it will trigger ENFORCE failure (exception).
In the current implementation, it may have carry-on impact. The mismatch ratio is equal to:
```
ratio = mismatched_ranges_from_t0_to_t1 / total_ranges_from_t0_to_t1
```
if the mismatched_ranges_from_t0 somehow increased a lot (bad value spike) at request N, the ratio will be much larger than the treshold.  Then it may take long util t2 to make the new ratio drops below the threshold. however, the requests between t1 and t2 may be all good requests, then it brings carry-on impact.
Instead, we would like to propose a new strategy, when exception happen at T1, we will clean up all the history counters for this bad feature, and make it a clean run for the next phase util the next exception. it then will get rid of the carry-on impact.
In this logic, we will clean up the counter based on the bad feature J.
more context: https://docs.google.com/document/d/1tYHISyiLf-PVKPVGlZRZ0iq2Hvog3g5BjCKMCDLZHHo/edit

Test Plan:
hardcode a much smaller threshold as 0.0001 to force trigger the exception, deploy on some hosts in prod tiers

```
EPHEMERAL_PACKAGE=d44a3de1305c3b4c30fd62bc354a1285 tw update fbcode/tupperware/config/admarket/sigrid/predictor/prod.tw tsp_cln/admarket/sigrid_predictor_v2_dh_t1_elastic_ha --tasks=100-199 --fast --force
```

## 1) totalRanges can be correct logged
```
I1018 20:37:09.012523  2074 gather_ranges_to_dense_op.h:69 req:00f00000001abbf1] In GatherRangesToDenseOp:
  Lifetime empty ranges for each feature is 12354.
  Lifetime mismatched ranges for each feature is 526.
  With a total of 87503 examples for each feature.
```

## 2) exception can be still triggered
```
E1018 21:08:42.007398   668 LoggingPredictorService.cpp:701 req:001000000013df51] getRequestPrecomputedDataOnePass failure on model 481948521_146: [enforce fail at gather_ranges_to_dense_op.h:215] std::max(totalRangesTemp, minObservation_) * maxMismatchedRatio_ >= mismatchedRangesTemp. 0.1 vs 1. Ratio of range length mismatch for feature at index 0 is 0.00813008 (1/123) which exceeds 1e-05. The incorrect lengths include: 15 (Error from operator:

```

Differential Revision: D50570811

malfet commented on caffe2/operators/gather_ranges_to_dense_op.h:
```diff
@@ -45,14 +45,19 @@ class GatherRangesToDenseOp final : public Operator<Context> {
   ~GatherRangesToDenseOp() noexcept override {
-    if (totalRanges_ > minObservation_) {
+    bool exceedMinObservation = false;
+    for (int i = 0; i < totalRanges_.size(); i++) {
+      exceedMinObservation |= totalRanges_[i] > minObservation_;
+    }
```
Use range loop if possible

```suggestion
    for (const auto range : totalRanges_) {
      exceedMinObservation |= range > minObservation_;
    }
```

dlmetamovefast replied:
will update. thanks

malfet commented on caffe2/operators/gather_ranges_to_dense_op.h:
```diff
@@ -173,42 +178,56 @@ class GatherRangesToDenseOp final : public Operator<Context> {
 
     // Check whether the empty and mismatch ratio exceeded the threshold.
-    totalRanges_ += batchSize;
+    for (const auto j : c10::irange(OutputSize())) {
+      totalRanges_[j] += batchSize;
+    }
```
Is `OutputSize()` should always be the same as totalRanges_.size why not use a range loop?
```suggestion
   assert(OputputSize() == totalRanges_.size());
    for (auto& range : totalRanges_)  {
      range += batchSize;
    }
```

dlmetamovefast replied:
should be, based on line 46 for (const auto i : c10::irange(OutputSize()))

malfet commented on caffe2/operators/gather_ranges_to_dense_op.h:
```diff
@@ -173,42 +178,56 @@ class GatherRangesToDenseOp final : public Operator<Context> {
-        CAFFE_ENFORCE_GE(
-            std::max(totalRanges_, minObservation_) * maxMismatchedRatio_,
-            mismatchedRanges_[j],
+		CAFFE_ENFORCE_GE(
```
Why indentation change here?

malfet reviewed (approved):
This is caffe2 change, so from OSS point of view it's almost completely irrelevant, but would be nice to have a unit test

dlmetamovefast committed:
[Resubmit][S372460 follow up] Reduce embedding feature validation failure carry-on impact (#111838)

Summary:

## Context
The embedding feature validation for GatherRangeToDense was added in the previous diff: D18031155. The logic will check the mismatchedranges or empty ranges in the whole model lifecycle, once the ratio exceeds some thresehold, it will trigger ENFORCE failure (exception).
In the current implementation, it may have carry-on impact. The mismatch ratio is equal to:
```
ratio = mismatched_ranges_from_t0_to_t1 / total_ranges_from_t0_to_t1
```
if the mismatched_ranges_from_t0 somehow increased a lot (bad value spike) at request N, the ratio will be much larger than the treshold.  Then it may take long util t2 to make the new ratio drops below the threshold. however, the requests between t1 and t2 may be all good requests, then it brings carry-on impact.
Instead, we would like to propose a new strategy, when exception happen at T1, we will clean up all the history counters for this bad feature, and make it a clean run for the next phase util the next exception. it then will get rid of the carry-on impact.
In this logic, we will clean up the counter based on the bad feature J.
more context: https://docs.google.com/document/d/1tYHISyiLf-PVKPVGlZRZ0iq2Hvog3g5BjCKMCDLZHHo/edit

Test Plan:
hardcode a much smaller threshold as 0.0001 to force trigger the exception, deploy on some hosts in prod tiers

```
EPHEMERAL_PACKAGE=d44a3de1305c3b4c30fd62bc354a1285 tw update fbcode/tupperware/config/admarket/sigrid/predictor/prod.tw tsp_cln/admarket/sigrid_predictor_v2_dh_t1_elastic_ha --tasks=100-199 --fast --force
```

## 1) totalRanges can be correct logged
```
I1018 20:37:09.012523  2074 gather_ranges_to_dense_op.h:69 req:00f00000001abbf1] In GatherRangesToDenseOp:
  Lifetime empty ranges for each feature is 12354.
  Lifetime mismatched ranges for each feature is 526.
  With a total of 87503 examples for each feature.
```

## 2) exception can be still triggered
```
E1018 21:08:42.007398   668 LoggingPredictorService.cpp:701 req:001000000013df51] getRequestPrecomputedDataOnePass failure on model 481948521_146: [enforce fail at gather_ranges_to_dense_op.h:215] std::max(totalRangesTemp, minObservation_) * maxMismatchedRatio_ >= mismatchedRangesTemp. 0.1 vs 1. Ratio of range length mismatch for feature at index 0 is 0.00813008 (1/123) which exceeds 1e-05. The incorrect lengths include: 15 (Error from operator:

```

Reviewed By: malfet, xychang

Differential Revision: D50570811
